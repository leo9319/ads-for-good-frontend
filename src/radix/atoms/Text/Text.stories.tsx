import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { Text, TextProps } from './Text';
import { Theme } from '@radix-ui/themes';

const meta: Meta<typeof Text> = {
  title: 'Library/Atoms/Text',
  component: Text,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  argTypes: {
    children: {
      control: 'text',
    },
    size: {
      description: 'Font size',
      control: {
        type: 'select',
      },
      options: [
        '1',
        '2',
        '3',
        '4',
        '5',
        '6',
        '7',
        '8',
        '9',
        '10',
        '11',
        '12',
        '13',
      ],
    },
    weight: {
      control: {
        type: 'select',
      },
      options: ['regular', 'semibold', 'bold'],
    },
    className: {
      control: 'text',
    },
  },
  args: {
    children: 'Sample Text',
    size: '3',
    weight: 'regular',
    as: 'span',
  },
};

export default meta;

type Story = StoryObj<TextProps>;

export const Default: Story = {
  args: {
    children: 'Default Text',
  },
};

export const LargeBoldText: Story = {
  args: {
    children: 'Large Bold Text',
    size: '7',
    weight: 'bold',
  },
};

export const TextWithHTML: Story = {
  args: {
    children:
      '<div><p>Hello, world!</p><span>React is great!</span><ul><li>List 1</li><li>List 2</li><li>List 3</li><li>List 4 <a href="#">some link1</a></li></ul></div>',
    size: '2',
    weight: 'regular',
  },
};
