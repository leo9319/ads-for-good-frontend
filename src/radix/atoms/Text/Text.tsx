import { TextSize, TextWeight } from '@internal/types/common/style';
import { Text as RadixText, TextProps as Props } from '@radix-ui/themes';
import { getFontStyles } from '@utils/elements/font';
import React, { Fragment, ReactElement } from 'react';
import { useContent } from '@utils/hooks';
import { Config } from '@utils/common/htmlToRadix';
import { hasTags } from '@utils/common/text';

export type TextProps = Omit<Props, 'size' | 'weight'> & {
  /**
   * The size of the text
   * @property
   */
  size?: TextSize | null;
  /**
   * The weight of the text
   * @property
   */
  weight?: TextWeight | null;
  /**
   * config for the text
   * @property
   */
  config?: TextProps;
};

/**
 * Text component is to apply the text style to the children element.
 *
 * @group Atoms
 * @category Radix Component
 *
 * @param {TextProps} props - The Text props
 * @returns {ReactElement} - Text Component
 */
export const Text = ({
  children,
  size = '3',
  weight = 'regular',
  className,
  config,
  ...props
}: TextProps): ReactElement => {
  const classes = getFontStyles(props, size, weight, className);
  const newProps = { className: classes, ...(props as Props) };

  const defaultTextConfig = {
    size,
    weight,
  };

  const textConfig: Config = {
    textProps: {
      ...(config ?? defaultTextConfig),
    },
  };

  const hasStringAsChildren = typeof children === 'string';
  const { id, newChildren } = useContent(
    hasStringAsChildren ? children : null,
    textConfig
  );

  if (hasStringAsChildren) {
    return (
      <RadixText
        id={id}
        {...newProps}
        asChild={props.asChild || hasTags(children)}
      >
        <Fragment>{newChildren}</Fragment>
      </RadixText>
    );
  }
  return <RadixText {...newProps}>{children}</RadixText>;
};

export default Text;
