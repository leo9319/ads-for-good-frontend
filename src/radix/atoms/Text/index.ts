export { Text } from './Text';
export { default } from './Text';
export type { TextProps } from './Text';
