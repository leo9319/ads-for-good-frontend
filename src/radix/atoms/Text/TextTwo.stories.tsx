import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { Text, TextProps } from './Text';
import { Theme } from '@radix-ui/themes';

const meta: Meta<typeof Text> = {
  title: 'Library/Atoms/Text',
  component: Text,
  parameters: {
    layout: 'centered',
  },
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  argTypes: {
    children: {
      control: 'text',
    },
    size: {
      control: {
        type: 'object',
      },
    },
    weight: {
      control: {
        type: 'object',
      },
    },
    className: {
      control: 'text',
    },
  },
  args: {
    children: 'Sample Text',
    size: '3',
    weight: 'regular',
    as: 'span',
  },
};

export default meta;

type Story = StoryObj<TextProps>;

export const BreakPointVariantText: Story = {
  args: {
    size: { initial: '1', xs: '2', sm: '3', md: '4', lg: '4', xl: '6' },
    weight: { initial: 'regular', sm: 'semibold', md: 'bold' },
    children: 'Custom Text with Breakpoint Variant Styles',
    className: 'custom-class',
  },
};
