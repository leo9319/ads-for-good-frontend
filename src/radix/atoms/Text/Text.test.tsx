import React from 'react';
import { render } from '@testing-library/react';
import Text from './Text';

describe('Text', () => {
  it('Should render successfully', () => {
    const { baseElement } = render(<Text>Text</Text>);
    expect(baseElement).toMatchSnapshot();
  });
});
