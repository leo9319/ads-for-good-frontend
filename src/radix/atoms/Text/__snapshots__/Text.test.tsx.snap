// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Should render successfully 1`] = `
<body>
  <div>
    <span
      class="rt-Text mocked-class-name typography-font-family-default"
      id=""
    >
      Text
    </span>
  </div>
</body>
`;
